### ISSUE TITLE
How to factorize code between kops and cluster api?

### ISSUE THREAD

**Issue by remyleone** [kubernetes-sigs/cluster-api#10572]
### What would you like to be added (User Story)?

As a cloud service provider I would like to have maximum support for all the present and future tools that Kubernetes sigs supports officially. Cluster API and Kops are both projects that requires maintenance and we would like to reuse as much as possible between those two projects. Is there best practices, exported packages/interfaces that could ease up both maintaining kops and cluster api support for a given cloud service provider?

### Detailed Description

I would like to have documentation about how to have reusable components between kops and cluster api controller. For those both tools there is a lot of overlap in the preparation of a cluster. Is there any documentation/recommendation about how to reuse code between those two projects?

I would like to have documentation about the core functions/interfaces that would need to be documented to permit maximum reusability between kops and cluster api.

### Anything else you would like to add?

https://github.com/kubernetes/kops/issues/16540

### Label(s) to be applied

/kind feature

**Comment by sbueringer** [comment]
I think we don't have anything like that. As far as I can tell just the documentation for CAPI is not very extensive. Mostly because nobody is volunteering to work on this documentation

**Comment by fabriziopandini** [comment]
The ask could make a really interesting discussion at sig level, but I don't think I don't think that reuse of code between CAPI and kops is in scope of these projects (disclaimer I'm not a kops contributor, but for Cluster API you can look at https://cluster-api.sigs.k8s.io/user/manifesto).

I'm leading towards closing this issue, because the Cluster API project currently lacks enough active contributors to adequately respond to all issues and PRs, and if this concern was never raised in the last few years, I don't think it is relevant for both communities and that we will have volunteers to get this addressed.

I will wait a few more days to see if someone will comment on this thread

**Comment by fabriziopandini** [comment]
chatted about this with @justinsb at SIG office hours yesterday, he is going to provide some recommendation to the corresponding issue in kops.

/close

**Comment by justinsb** [comment]
I added a comment here: https://github.com/kubernetes/kops/issues/16540#issuecomment-2138912300

TLDR I do think that every surface ends up implementing a reconciliation layer over the REST API, and that potentially we could reuse that.  I also think that we could look at building cloud operators like KCC/ACK/ASO and investigate how to use those in cloud-providers / cluster-api / kOps, but that's more speculative.
